msglist: Support viewing who reacted to a message

chrisbobbe opened the pull request:
Fixes #740.

Followup:

- #1747

Notable changes from previous revision:

- "See who reacted" button omitted in action sheet when no reactions
- Emojis centered horizontally when few
- Emojis scroll into view when tapped
- Reasonable UI with VoiceOver (at least in my testing)
- Wrote some basic tests

-----

### Screenshots

I've included some screenshots where there are
- many emoji
- many users

and in those, I set the scroll position such that you can see the shadow effect.

| Light | Dark |
| --- | --- |
| <img width="559" height="1062" alt="image" src="https://github.com/user-attachments/assets/0b05a7a5-6de2-4710-976a-a96f6b942b28" /> | <img width="559" height="1062" alt="image" src="https://github.com/user-attachments/assets/801c05f8-e106-4ab7-a2f6-7d793070bcb6" /> |
| <img width="559" height="1062" alt="image" src="https://github.com/user-attachments/assets/303b49ac-b6f8-4c45-bb0f-43adf91bf1f9" /> | <img width="559" height="1062" alt="image" src="https://github.com/user-attachments/assets/4ddff0ad-7ac6-405f-8bbb-c38ad6f6de14" /> |
| <img width="559" height="1062" alt="image" src="https://github.com/user-attachments/assets/d847f81f-b419-4486-9b8a-1443e3a39f35" /> | <img width="559" height="1062" alt="image" src="https://github.com/user-attachments/assets/968525cc-ba06-4554-ad0d-7e0ccfdac593" /> |
| <img width="559" height="1062" alt="image" src="https://github.com/user-attachments/assets/8ba1bc38-8c2b-47ab-8c85-409c71433e4f" /> | <img width="559" height="1062" alt="image" src="https://github.com/user-attachments/assets/e24536d7-b211-46e0-8749-8e70c5a1a95e" /> |

Scroll-on-select animation:

![Jul-31-2025 17-27-36](https://github.com/user-attachments/assets/ae882893-7f34-48a5-8b7f-4a6e12294c56)

chrisbobbe commented:
Thanks! This is now ready for review, and I've updated the issue description with changes since the draft.

gnprice commented:
Thanks! Would you also post a few screenshots?

chrisbobbe commented:
Sure! Done.

gnprice commented on lib/widgets/emoji_reaction.dart:
```diff
@@ -614,3 +627,411 @@ class EmojiPickerListEntry extends StatelessWidget {
+
+  void _onPressed() {
+    // Dismiss the action sheet.
+    Navigator.pop(pageContext);
+
+    Navigator.push(pageContext,
+      ProfilePage.buildRoute(context: pageContext, userId: userId));
```
I think the reason this works even though this `pageContext` isn't really a page context is that there's no `await` here.

The reason our action-sheet buttons often need a real page context is that they'll dismiss the action sheet, then do something that takes time like a network request, and then after that completes they need to use a context to do something with the result, or to show an error. So if the request takes longer than the few hundred ms of the dismiss animation, the action sheet's own context may be unmounted by that point.

For just synchronously acting on the navigation like this, the widget's own context is fine.

gnprice commented on lib/widgets/emoji_reaction.dart:
```diff
@@ -614,3 +627,411 @@ class EmojiPickerListEntry extends StatelessWidget {
+      onTap: _onPressed,
+      splashFactory: NoSplash.splashFactory,
+      overlayColor: WidgetStateColor.resolveWith((states) =>
+        states.any((e) => e == WidgetState.pressed)
```
```suggestion
        states.contains(WidgetState.pressed)
```

Looks like it's a Set, so that should be equivalent (and potentially faster)

gnprice reviewed (commented):
Thanks!

Generally this looks good. Here's a full review except the tests.

chrisbobbe commented on lib/widgets/emoji_reaction.dart:
```diff
@@ -614,3 +627,411 @@ class EmojiPickerListEntry extends StatelessWidget {
+          //    to the underlying Scrollable to remove an unwanted node
+          //    in accessibility focus traversal.
+          scrollDirection: Axis.horizontal,
+          physics: ClampingScrollPhysics(),
```
Ah yeah I think the default is fine actually :)

I was reading a tab-bar implementation (when doing the Semantics work for this PR) and I think it opted out of the "bouncing" behavior. But I'm not sure this is really a "tab bar" in the same sense as that code I was reading.

chrisbobbe commented:
Thanks for the review! Revision pushed. I noticed on reaction chips that the tooltip-on-long-press was sometimes winning out over the show-view-reactions-sheet, so I added a commit to get rid of the tooltip, with a test change for some tests that had been relying on it.

I also added a tweak to the scroll-into-view logic for the horizontal list of emojis, as a separate commit on top.

gnprice commented on lib/widgets/emoji_reaction.dart:
```diff
@@ -595,3 +635,446 @@ class EmojiPickerListEntry extends StatelessWidget {
+
+  /// A [double] between 0.0 and 1.0 for an emoji's position in the list.
+  ///
+  /// When auto-scrolling an emoji into view,
+  /// this is where the scroll position will land
+  /// (the min- and max- scroll extent lerped to this value).
+  double _emojiItemPosition(int index, int aggregatedLength) {
```
lerped "at"?

Or some other/further rewording. I'd read "lerped to X" as meaning that X was the second argument of a `lerp` method — conceptually the arguments are start, end, and time.

gnprice commented on lib/widgets/emoji_reaction.dart:
```diff
@@ -595,3 +635,446 @@ class EmojiPickerListEntry extends StatelessWidget {
+  ///   any item is tapped, because each item has a different [position].)
+  void _scrollIntoView(BuildContext context) {
+    final scrollPosition = Scrollable.of(context, axis: Axis.horizontal).position;
+    final destination = lerpDouble(
+      scrollPosition.minScrollExtent,
+      scrollPosition.maxScrollExtent,
+      position);
+    if (destination == null) return; // TODO(log)
```
This condition is impossible, right? (From the definition of lerpDouble.) So just use `!`.

chrisbobbe commented:
Thanks! Revision pushed.

chrisbobbe commented on lib/widgets/emoji_reaction.dart:
```diff
@@ -595,3 +641,447 @@ class EmojiPickerListEntry extends StatelessWidget {
+    }
+
+    Widget result = SizedBox(
+      height: 400,
```
Oh oops, I meant to leave a `TODO(design) tune` here (I think I had it in an earlier revision)—will put it back in my next revision

gnprice commented on lib/widgets/emoji_reaction.dart:
```diff
@@ -614,3 +627,411 @@ class EmojiPickerListEntry extends StatelessWidget {
+    return SizedBox(
+      width: double.infinity,
+      child: Column(
+        mainAxisSize: MainAxisSize.min,
```
Very helpful, thanks! OK, so an earlier version of this PR (as seen in v30.0.262) had "stretch" alignment. That makes the children wide, including the header; and then I think the Row in the header, with defaults MainAxisSize.max and MainAxisAlignment.start, ends up putting the items at the start.

The revised version looks good.

chrisbobbe commented:
Done, thanks!
